In [ ]:
from google import genai
from google.colab import userdata

api_key = userdata.get('Gemini_API_Key')
client = genai.Client(api_key=api_key)

In [ ]:
# Função auxiliar para fazer chamadas à API do Gemini
def obter_resposta_llm(texto_prompt, modelo="gemini-3.6-flash"):
    if not client:
        return "Erro: O cliente da API não foi inicializado."

    try:
        resposta = client.models.generate_content(
            model=modelo,
            contents=texto_prompt,
        )
        return resposta.text
    except Exception as e:
        return f"Erro na comunicação com a API do Gemini: {e}"


In [ ]:
problema='''Problema: Uma loja aumenta o preço de um produto em 20% e depois
dá um desconto de 20% sobre o novo preço. O preço final é igual,
maior ou menor que o original? '''

solucao_inicial=obter_resposta_llm(problema)
print("\n---Processo de Reflexão---")
print(f'Solução inicial: {solucao_inicial}')



---Processo de Reflexão---
Solução inicial: O preço final é **menor** que o preço original.

### Explicação simples (com exemplo):

Imagine que o produto custava **R$ 100,00**:

1. **Aumento de 20%:** 
   * 20% de R$ 100,00 é R$ 20,00.
   * O novo preço passa a ser **R$ 120,00**.

2. **Desconto de 20% sobre o novo preço:**
   * Agora, os 20% de desconto são calculados sobre R$ 120,00 (e não mais sobre R$ 100,00).
   * 20% de R$ 120,00 é **R$ 24,00**.
   * Preço final: R$ 120,00 - R$ 24,00 = **R$ 96,00**.

**Resultado:** O preço final (R$ 96,00) ficou **4% menor** do que o preço original (R$ 100,00).

Isso acontece porque o desconto foi aplicado sobre um valor maior (R$ 120,00), gerando uma redução em reais maior do que o aumento inicial.


In [ ]:
prompt_de_critica=f'''Revise a {solucao_inicial}, a algum erro? A solução está correta?

'''

In [ ]:
if client:
  critica=obter_resposta_llm(prompt_de_critica)
  print(f'Critíca: {critica}')

Critíca: **Não há nenhum erro. A solução está 100% correta!**

Tanto a matemática quanto a lógica e a Didática da explicação estão perfeitas. 

### Por que o seu texto está ótimo:

1. **A matemática está exata:** 
   * $100 \times 1,20 = 120$
   * $120 \times 0,80 = 96$
   * A diferença de R$ 100 para R$ 96 é exatamente **4% de redução**.

2. **A didática está perfeita:** Usar o valor inicial de **R$ 100,00** é a melhor escolha pedagógica para explicar porcentagem, pois facilita para o leitor visualizar a relação direta entre o valor em reais e a porcentagem.

3. **A conclusão é precisa:** A frase final (*"Isso acontece porque o desconto foi aplicado sobre um valor maior..."*) resolve a dúvida conceitual mais comum sobre esse tema (o mito de que +20% e -20% se anulam).

Seu texto está claro, bem formatado e pronto para ser usado!


In [ ]:
prompt_final_refinado=f'''Com base no {solucao_inicial} e na {critica}, reecreva a solução do {problema}
'''

In [ ]:
if client:
  solucao_final=obter_resposta_llm(prompt_final_refinado)
  print(f'Solução Final: {solucao_final}')


Solução Final: Aqui está a solução reescrita de forma clara, didática e estruturada, pronta para ser apresentada:

---

# Solução do Problema: Aumento e Desconto Sucessivos

**Resposta:** O preço final é **menor** do que o preço original. Ele fica **4% mais barato**.

---

### Explicação passo a passo (Exemplo Prático)

Para visualizar facilmente, imagine que o produto custava originalmente **R$ 100,00**:

1. **1º Passo: Aumento de 20%**
   * Calculamos 20% sobre o preço original (R$ 100,00) = **R$ 20,00**.
   * Novo preço: R$ 100,00 + R$ 20,00 = **R$ 120,00**.

2. **2º Passo: Desconto de 20% sobre o novo preço**
   * O desconto agora é calculado sobre o valor atualizado (**R$ 120,00**), e não mais sobre o valor inicial.
   * Calculamos 20% de R$ 120,00 = **R$ 24,00**.
   * Preço final: R$ 120,00 - R$ 24,00 = **R$ 96,00**.

---

### Por que isso acontece? (A Lógica por trás)

Embora as porcentagens pareçam se anular (+20% e -20%), **a base de cálculo mudou**:
* O aumento foi de 20% sob

Solução Final: Aqui está a solução reescrita de forma clara, didática e estruturada, pronta para ser apresentada:

---

# Solução do Problema: Aumento e Desconto Sucessivos

**Resposta:** O preço final é **menor** do que o preço original. Ele fica **4% mais barato**.

---

### Explicação passo a passo (Exemplo Prático)

Para visualizar facilmente, imagine que o produto custava originalmente **R\$ 100,00**:

1. **1º Passo: Aumento de 20%**

    - Calculamos 20% sobre o preço original (R\$ 100,00) = **R\$ 20,00**
    - Novo preço: R\$ 100,00 + R\$ 20,00 = **R\$ 120,00**

2. **2º Passo: Desconto de 20% sobre o novo preço**

    - O desconto agora é calculado sobre o valor atualizado (**R\$ 120,00**), e não mais sobre o valor inicial
    - Calculamos 20% de R\$ 120,00 = **R\$ 24,00**
    - Preço final: R\$ 120,00 - R\$ 24,00 = **R\$ 96,00**

---

### Por que isso acontece? (A Lógica por trás)

Embora as porcentagens pareçam se anular (+20% e -20%), **a base de cálculo mudou**:

- O aumento foi de 20% sobre um valor menor (R\$ 100,00), gerando uma alta de **R\$ 20,00**
- O desconto foi de 20% sobre um valor maior (R\$ 120,00), gerando uma redução de **R\$ 24,00**

Como o desconto em dinheiro foi maior do que o aumento inicial, o produto acabou custando menos.

---

### Demonstração Matemática Direta

$$\text{Preço Final} = \text{Preço Inicial} \times 1{,}20 \text{ (aumento)} \times 0{,}80 \text{ (desconto)}$$

$$100 \times 1{,}20 \times 0{,}80 = \mathbf{96}$$

**Resultado Final:** O preço final equivale a 96% do preço original, ou seja, houve uma **redução total de 4%**.